job-info: support LRU cache mapping job id -> owner

chu11 opened the pull request:
per discussion in #3521 

consideration: should I drop the commit for the "always-lookup-job-owner" option?  It's very specific to testing this LRU support, and isn't something I even bothered to add a test into sharness for.  It really can't be used for anything else.  I'm leaning towards removing that commit, but thought I'd leave in for the time being.

garlick reviewed (commented):
LGTM, just a couple minor suggestions for you to consider.

chu11 commented on src/modules/job-info/allow.c:
```diff
@@ -60,11 +60,30 @@ static int eventlog_get_userid (struct info_ctx *ctx, const char *s,
     return rv;
 }
 
+static void store_lru (struct info_ctx *ctx, flux_jobid_t id, int userid)
+{
+    char id_buf[64];
+    char *userid_ptr = NULL;
+
+    snprintf (id_buf, 64, "%ju", (uintmax_t)id);
+
+    if (asprintf (&userid_ptr, "%d", userid) < 0)
+        return;
+
+    if (lru_cache_put (ctx->owner_lru, id_buf, userid_ptr) < 0) {
+        if (errno != EEXIST)
+            flux_log_error (ctx->h, "%s: lru_cache_put", __FUNCTION__);
+        free (userid_ptr);
+        return;
+    }
+    return;
+}
+
```
I think you mean the `userid`, but yup, makes sense

chu11 commented:
rebased & re-pushed with fixes per @garlick comments above

I removed the "always-lookup-job-owner" commit b/c it is so specific to testing this feature, didn't feel it was necessary to be merged into master.

Re-ran on `fluke` just to make sure the minor performance benefit was still there.

w/ LRU

```
  8192 jobs:   8192 submitted,      0 running, 8192 completed
number of jobs: 8192
script runtime: 245.219s
job runtime:    243.933s
throughput:     33.6 job/s (script:  33.4 job/s)
```

w/o LRU

```
  8192 jobs:   8192 submitted,      0 running, 8192 completed
number of jobs: 8192
script runtime: 247.376s
job runtime:    246.047s
throughput:     33.3 job/s (script:  33.1 job/s)
```

garlick reviewed (approved):
Looks good @chu11.  Thanks!  Feel free to set MWP when you're ready.

chu11 commented:
@garlick, thanks!

@mergifyio rebase

grondo commented:
> Mergifyio rebase

Just fyi, merfiy will auto-rebase before merging when MWP is set, so no need to request it manually. :+1:
